cmd/geth/chaincmd.go

```diff
@@ -595,3 +608,67 @@ func hashish(x string) bool {
 	_, err := strconv.Atoi(x)
 	return err != nil
 }
+
+func pruneHistory(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	// Open the chain database
+	chain, chaindb := utils.MakeChain(ctx, stack, false)
+	defer chaindb.Close()
+	defer chain.Stop()
+
+	// Determine the prune point. This will be the first PoS block.
+	prunePoint, ok := ethconfig.HistoryPrunePoints[chain.Genesis().Hash()]
+	if !ok || prunePoint == nil {
+		return errors.New("prune point not found")
+	}
+	var (
+		mergeBlock     = prunePoint.BlockNumber
+		mergeBlockHash = prunePoint.BlockHash.Hex()
+	)
+
+	// Check we're far enough past merge to ensure all data is in freezer
+	currentHeader := chain.CurrentHeader()
+	if currentHeader == nil {
+		return errors.New("current header not found")
+	}
+	if currentHeader.Number.Uint64() < mergeBlock+params.FullImmutabilityThreshold {
+		return fmt.Errorf("chain not far enough past merge block, need %d more blocks",
+			mergeBlock+params.FullImmutabilityThreshold-currentHeader.Number.Uint64())
+	}
+
+	// Double-check the prune block in db has the expected hash.
+	hash := rawdb.ReadCanonicalHash(chaindb, mergeBlock)
+	if hash != common.HexToHash(mergeBlockHash) {
+		return fmt.Errorf("merge block hash mismatch: got %s, want %s", hash.Hex(), mergeBlockHash)
+	}
+
+	txlookupTail := rawdb.ReadTxIndexTail(chaindb)
+
+	log.Info("Starting chain pruning",
+		"currentHeight", currentHeader.Number,
+		"mergeBlock", mergeBlock,
+		"mergeBlockHash", mergeBlockHash)
+
+	start := time.Now()
+
+	// First prune the transaction lookup index as
+	// it requires the block bodies.
+	if txlookupTail != nil && *txlookupTail < mergeBlock {
+		rawdb.UnindexTransactions(chaindb, *txlookupTail, mergeBlock, nil, false)
```

Comment:
This takes forever to run. We discussed offline, and it seems it would be best to either:

- Implement a routine that scans the whole index in hash order, and removes entries that have a block older than the cutoff number.
- Remove it from here entirely, and leave it to the tx indexer to remove the broken entries. However, if we do this, we'd need to deal with potentially outdated entries during API call processing.